### DISCUSSION TITLE
Need help making vike redirects working with Vercel deployments

### DISCUSSION THREAD

**Discussion by ndavd** [vikejs/vike#1244] [Discussion] [answered]
So I am using vike redirects (https://vike.dev/redirects) which work fine in dev. But on Vercel they don't work at all (just get a blank page). Also tried using `vercel.json` redirects instead but then those don't work in dev and also don't support `mailto` URLs.

---

Here's the minimal Vercel setup I'm using (based on https://vike.dev/vercel#api-route):
- `vercel.json`
```json
{
  "rewrites": [
    {
      "source": "/(.*)",
      "destination": "/api/ssr.js"
    }
  ]
}
```
- `api/ssr.js`
```js
import { renderPage } from 'vike/server'

// We use JSDoc instead of TypeScript because Vercel seems buggy with /api/**/*.ts files

/**
 * @param {import('@vercel/node').VercelRequest} req
 * @param {import('@vercel/node').VercelResponse} res
 */
export default async function handler(req, res) {
  const { url } = req
  console.log('Request to url:', url)
  if (url === undefined) throw new Error('req.url is undefined')

  const pageContextInit = { urlOriginal: url }
  const pageContext = await renderPage(pageContextInit)
  const { httpResponse } = pageContext
  console.log('httpResponse', !!httpResponse)

  if (!httpResponse) {
    res.statusCode = 200
    res.end()
    return
  }

  const { body, statusCode, contentType } = httpResponse
  res.statusCode = statusCode
  res.setHeader('content-type', contentType)
  res.end(body)
}
```

**Comment by brillout** [comment]
That's quite unexpected because what Vike does is adding the HTTP redirect header to the HTTP response. I doubt Vercel removes the HTTP redirect header?

**Reply by ndavd** [reply]
Hmm, if I share the URL to one of said routes would that help?

**Reply by brillout** [reply, marked as answer]
Actually, you're using a deprecated `renderPage()` integration. Make sure to update your `renderPage()` integration to latest version (you should be getting a deprecation warning pointing you to https://vike.dev/migration/0.4.134#vercel).

**Reply by brillout** [reply]
@ndavd Btw. would your company be up for [sponsoring](https://github.com/sponsors/brillout)?

**Reply by ndavd** [reply]
Oh awesome thanks! Also, I noticed that here I can't also redirect to a `mailto` URL? `Error: [vike][Wrong Usage][vite.config.js > vike({ redirects })] Invalid redirection target URL mailto:foo@bar.test: the target URL should start with /, http://, https://, or be *`

**Reply by ndavd** [reply]
> @ndavd Btw. would your company be up for [sponsoring](https://github.com/sponsors/brillout)?

Hey, I don't know. I just found Vike yesterday and I'm using it for my personal website (https://ndavd.com)

**Reply by brillout** [reply]
> redirect to a `mailto` URL

Curious: what's your use case?

I'll have a look if `mailto:` can be whitelisted later today.

**Reply by ndavd** [reply]
> Curious: what's your use case?

I like to have redirects like `https://ndavd.com/mail` -> `mailto:email@ndavd.com` (take a look at my links over on https://ndavd.com)

> I'll have a look if `mailto:` can be whitelisted later today.

That would be awesome!! From my side, I am loving Vike (I became a Vite fan a few months ago) and will be using it moving forward on my projects. I'll also advertise for it on my personal website as I usually do with the technologies I use the most :)

**Reply by ndavd** [reply]
Also @brillout if you do end up getting `mailto:` whitelisted today, how can I use it, need to wait until next Vite version?

EDIT: Additional question, why do those urls that get accepted into the request have to be whitelisted at all? Would it be bad if they weren't? Because I can think of some other types of url that might have some usecases, `ipfs://`, `magnet:` (torrents also have a lot of legitimate usecases such as distribution of Linux isos and other heavy files), etc.

**Reply by brillout** [reply]
> I'll have a look if `mailto:` can be whitelisted

I just had a look at it and, unfortunately, it turns out that it isn't as trivial as I initially thought.

But feel free to create a new feature request for it (if you do then make sure to describe a rationale & motivation so that I can prioritize accordingly).

> `https://ndavd.com/mail` -> `mailto:email@ndavd.com`

Why not pointing the link directly to `mailto:*` instead of going over an HTTP redirect?

> some other types of url that might have some usecases

Other than your use case, I actually cannot see other use cases &mdash; feel free to prove me wrong 😉

I'm glad you're enjoying Vike 🙂

**Reply by ndavd** [reply]
> Why not pointing the link directly to `mailto:*` instead of going over an HTTP redirect?

Because I want to have easy routes always pointing to the up-to-date contacts/socials:
`/li` always points to my LinkedIn, `/x` always points to my Twitter, `/mail` always redirects to the email I'm currently using

**Reply by brillout** [reply]
Makes sense.

Feature request welcome :)

**Reply by ndavd** [reply]
> Feature request welcome :)

Done https://github.com/vikejs/vike/issues/1245

Thanks for the support and prompt replies. Updating `api/ssr.js` fixed the issue for me 👍🏻
